src/V3Timing.cpp

```diff
@@ -0,0 +1,785 @@
+        m_cfuncp = nodep;
+        iterateChildren(nodep);
+        if (nodep->isCoroutine()
+            && !nodep->user1()) {  // If coroutine but no awaits (e.g. a class method that
+                                   // overrides/is overriden by a suspendable, but doesn't have any
+                                   // awaits itself)
+            // Add a co_return at the end so the compiler doesn't complain
+            nodep->addStmtsp(new AstCStmt{nodep->fileline(), "co_return;\n"});
+            nodep->user1(true);  // So we won't add another
+        }
+        m_cfuncp = nullptr;
+        m_lastTimingSensp = nullptr;
+    }
+    virtual void visit(AstNodeProcedure* nodep) override {
+        iterateChildren(nodep);
+        m_lastTimingSensp = nullptr;
+    }
+    virtual void visit(AstNodeVarRef* nodep) override {
+        if (m_lastTimingSensp && nodep->access().isWriteOrRW()) {
+            auto*& senTreep = m_timingKit.varToSenTreep[nodep->varScopep()];
+            if (senTreep) {
+                if (senTreep->user1()) {
+                    // Clone it so we don't modify a global sentree
+                    senTreep = senTreep->cloneTree(false);
+                    senTreep->user1p(nullptr);
+                }
+                senTreep->addSensesp(m_lastTimingSensp->sensesp()->cloneTree(true));
+            } else {
+                senTreep = m_lastTimingSensp;
+            }
+        }
+    }
+
+    //--------------------
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+public:
+    // CONSTRUCTORS
+    explicit TimingToCStmtsVisitor(AstNetlist* nodep, V3Timing::TimingKit& evalStmts)
+        : m_scopeTopp{nodep->topScopep()->scopep()}
+        , m_timingKit{evalStmts} {
+        nodep->typeTablep()->addTypesp(m_forkDtp);
+        nodep->typeTablep()->addTypesp(m_dlySchedDtp);
+        nodep->typeTablep()->addTypesp(m_trigSchedDtp);
+
+        // Delay scheduler has to be accessible from top
+        m_dlySchedp->varp()->sigPublic(true);
+
+        // Create a delay scheduler resume() call
+        auto* const flp = m_scopeTopp->fileline();
+        auto* const resumep
+            = new AstCMethodHard{flp, new AstVarRef{flp, m_dlySchedp, VAccess::WRITE}, "resume"};
+        resumep->dtypeSetVoid();
+        resumep->statement(true);
+        m_timingKit.preStmts = VN_AS(AstNode::addNext(m_timingKit.preStmts, resumep), NodeStmt);
+
+        // Run the visitor
+        iterate(nodep);
+
+        // Uniquify new sentrees and create actives for them
+        SenTreeFinder finder(nodep);
+        for (auto& p : evalStmts.varToSenTreep) {
+            if (p.second->user1()) continue;  // Not a new sentree
+            auto* const sensesp = finder.getSenTree(p.second);
+            p.second->deleteTree();
+            p.second = sensesp;
+            // Create an active so that a trigger gets generated for the sentree
+            createActive(sensesp);
+        }
```

Comment:
Instead of this, just change the V3Order `externalDomains` callback so you it returns a set of AstSenTrees* (a set is ok here, as V3Order will sort the sen list), then in V3Order, just combine all terms. There will likely be extra triggers needed for some stuff (I know #3429 will require some), so we might as well handle this generically in V3Order. I can change the API here for you if you prefer.

Reply:
Fixed